src/main/java/de/triology/cas/logout/LogoutUriEnabledLogoutManagerImpl.java

```diff
@@ -0,0 +1,200 @@
+import org.apache.commons.codec.binary.Base64;
+import org.jasig.cas.authentication.principal.Service;
+import org.jasig.cas.authentication.principal.SingleLogoutService;
+import org.jasig.cas.logout.LogoutMessageCreator;
+import org.jasig.cas.logout.LogoutRequest;
+import org.jasig.cas.logout.LogoutRequestStatus;
+import org.jasig.cas.services.LogoutType;
+import org.jasig.cas.services.RegisteredService;
+import org.jasig.cas.services.ServicesManager;
+import org.jasig.cas.ticket.TicketGrantingTicket;
+import org.jasig.cas.util.HttpClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This logout manager handles the Single Log Out process.
+ *
+ * @author Jerome Leleu
+ * @since 4.0.0
+ */
+public final class LogoutUriEnabledLogoutManagerImpl implements org.jasig.cas.logout.LogoutManager {
+
+    /** The logger. */
+    private static final Logger LOGGER = LoggerFactory.getLogger(LogoutUriEnabledLogoutManagerImpl.class);
+
+    /** ASCII character set. */
+    private static final Charset ASCII = Charset.forName("ASCII");
+
+    /** The services manager. */
+    @NotNull
+    private final ServicesManager servicesManager;
+
+    /** An HTTP client. */
+    @NotNull
+    private final HttpClient httpClient;
+
+    @NotNull
+    private final LogoutMessageCreator logoutMessageBuilder;
+
+    /** Whether single sign out is disabled or not. */
+    private boolean disableSingleSignOut = false;
+
+    /**
+     * Build the logout manager.
+     * @param servicesManager the services manager.
+     * @param httpClient an HTTP client.
+     * @param logoutMessageBuilder the builder to construct logout messages.
+     */
+    public LogoutUriEnabledLogoutManagerImpl(final ServicesManager servicesManager, final HttpClient httpClient,
+                                             final LogoutMessageCreator logoutMessageBuilder) {
+        this.servicesManager = servicesManager;
+        this.httpClient = httpClient;
+        this.logoutMessageBuilder = logoutMessageBuilder;
+    }
+
+    /**
+     * Perform a back channel logout for a given ticket granting ticket and returns all the logout requests.
+     *
+     * @param ticket a given ticket granting ticket.
+     * @return all logout requests.
+     */
+    @Override
+    public List<LogoutRequest> performLogout(final TicketGrantingTicket ticket) {
+        final Map<String, Service> services;
+        // synchronize the retrieval of the services and their cleaning for the TGT
+        // to avoid concurrent logout mess ups
+        synchronized (ticket) {
+            services = ticket.getServices();
+            ticket.removeAllServices();
+        }
+        ticket.markTicketExpired();
+
+        final List<LogoutRequest> logoutRequests = new ArrayList<>();
+        // if SLO is not disabled
+        if (!disableSingleSignOut) {
+            // through all services
+            for (final String ticketId : services.keySet()) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Iterate over the "entrySet" instead of the "keySet". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AS2864)